python/pyspark/sql/tests.py

```diff
@@ -4279,6 +4272,386 @@ def test_unsupported_types(self):
+        self.assertPandasEqual(expected5.toPandas(), result5.toPandas())
+
+    def test_multiple(self):
+        from pyspark.sql.functions import col, lit, sum, mean
+
+        df = self.data
+        mean_udf = self.mean_udf
+        sum_udf = self.sum_udf
+        weighted_mean_udf = self.weighted_mean_udf
+
+        result1 = (df.groupBy('id')
+                   .agg(mean_udf(df.v),
+                        sum_udf(df.v),
+                        weighted_mean_udf(df.v, df.w))
+                   .sort('id')
+                   .toPandas())
+
+        expected1 = (df.groupBy('id')
+                     .agg(mean(df.v).alias('mean_udf(v)'),
+                          sum(df.v).alias('sum_udf(v)'),
+                          mean(df.v).alias('weighted_mean_udf(v, w)'))
+                     .sort('id')
+                     .toPandas())
+
+        result2 = (df.groupBy('id', 'v')
+                   .agg(mean_udf(df.v),
+                        sum_udf(df.id))
+                   .sort('id', 'v')
+                   .toPandas())
+
+        expected2 = (df.groupBy('id', 'v')
+                     .agg(mean_udf(df.v).alias('mean_udf(v)'),
+                          sum_udf(df.id).alias('sum_udf(id)'))
+                     .sort('id', 'v')
+                     .toPandas())
+
+        self.assertPandasEqual(expected1, result1)
+        self.assertPandasEqual(expected2, result2)
+
+    def test_complex_grouping(self):
+        from pyspark.sql.functions import lit, sum
+
+        df = self.data
+        sum_udf = self.sum_udf
+        plus_one = self.plus_one
+        plus_two = self.plus_two
+
+        result1 = df.groupby(df.id + 1).agg(sum_udf(df.v))
+        expected1 = df.groupby(df.id + 1).agg(sum(df.v).alias('sum_udf(v)'))
+
+        result2 = df.groupby().agg(sum_udf(df.v))
+        expected2 = df.groupby().agg(sum(df.v).alias('sum_udf(v)'))
+
+        result3 = df.groupby(df.id, df.v % 2).agg(sum_udf(df.v))
+        expected3 = df.groupby(df.id, df.v % 2).agg(sum(df.v).alias('sum_udf(v)'))
+
+        result4 = df.groupby(plus_one(df.id)).agg(sum_udf(df.v))
+        expected4 = df.groupby(plus_one(df.id)).agg(sum(df.v).alias('sum_udf(v)'))
+
+        result5 = df.groupby(plus_two(df.id)).agg(sum_udf(df.v))
+        expected5 = df.groupby(plus_two(df.id)).agg(sum(df.v).alias('sum_udf(v)'))
+
+        result6 = df.groupby(df.id, plus_one(df.id)).agg(sum_udf(df.v))
+        expected6 = df.groupby(df.id, plus_one(df.id)).agg(sum(df.v).alias('sum_udf(v)'))
+
+        self.assertPandasEqual(expected1.toPandas(), result1.toPandas())
+        self.assertPandasEqual(expected2.toPandas(), result2.toPandas())
+        self.assertPandasEqual(expected3.toPandas(), result3.toPandas())
+        self.assertPandasEqual(expected4.toPandas(), result4.toPandas())
+        self.assertPandasEqual(expected5.toPandas(), result5.toPandas())
+        self.assertPandasEqual(expected6.toPandas(), result6.toPandas())
+
+    def test_complex_expression(self):
```

Comment:
`test_complex_mixed_udfs` ?

Reply:
I am leaning towards keeping the existing name. This doesn't just test mixed udfs, but also mixed udf and sql expressions. I added some comments in the test to be more clear. What do you think?